### PR TITLE
[KG] Multi-GPU kg training in pytorch

### PR DESCRIPTION
**Pull request by ZenoTan**
## Description

Use shared graph store to enable multi-gpu training. Embeddings are on CPUs, and processes pass them to GPUs to train.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

## Changes

Training uses subgraphs.
Shared graph store now works with test part.

FB15k, distmult, 1024 * 400

1 GPU:
500 steps take 10.092 seconds

2 GPU:
500 steps take 12.295 seconds

4 GPU:
500 steps take 11.920 seconds

8 GPU:
500 steps take 13.005 seconds

### CODE REVIEW

File: apps/kg/dataloader/sampler.py [hunk 1]
```diff
@@ -172,7 +172,7 @@ def __init__(self, g, edges, batch_size, neg_sample_size, mode, num_workers):
                                    num_workers=num_workers,
                                    shuffle=False,
                                    exclude_positive=False,
-                                   relations=g.edata['id'],
+                                   relations=g.edges[:].data['id'],
```
Comment (zheng-da) [inline apps/kg/dataloader/sampler.py]: what is the difference?

Comment (ZenoTan) [reply apps/kg/dataloader/sampler.py]: Graph store does not support edata and ndata (all edges/nodes) for now, so I just hacked the code like that.

File: apps/kg/models/general_models.py [hunk 2]
```diff
@@ -122,7 +125,9 @@ def forward_test(self, pos_g, neg_g, logs, gpu_id=-1):
 
         # We need to filter the positive edges in the negative graph.
         filter_bias = reshape(neg_g.edata['bias'], batch_size, -1)
-        if self.args.gpu >= 0:
+        if gpu_id != -1:
+            filter_bias = cuda(filter_bias, gpu_id)
+        elif self.args.gpu >= 0:
```
[the remark this answers is not in the document]
Comment (ZenoTan) [reply apps/kg/models/general_models.py]: Used separately for multi gpu and single gpu.

[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/models/general_models.py]: when it's multi-gpu training, `args.gpu` indicates the number of GPUs?

File: apps/kg/train.py [hunk 3]
```diff
@@ -127,8 +129,83 @@ def get_logger(args):
...
 
+def run_server(num_worker, graph, etype_id):
+    g = dgl.contrib.graph_store.create_graph_store_server(graph, "Test", "shared_mem",
+                    num_worker, False, edge_dir='in')
+    g.ndata['id'] = F.arange(0, graph.number_of_nodes())
+    g.edata['id'] = F.tensor(etype_id, F.int64)
+    g.run()
 
-def run(args, logger):
+def run_torch(args, logger):
```
Comment (zheng-da) [inline apps/kg/train.py]: why is this code pytorch specific?

Comment (zheng-da) [reply apps/kg/train.py]: does the code not work in MXNet?

File: apps/kg/models/general_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZenoTan) [reply apps/kg/models/general_models.py]: Now I've fixed it to indicate gpu list.

[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/models/general_models.py]: if `args.gpu` is a list, what this comparison means?

File: apps/kg/train.py [hunk 4]
```diff
@@ -60,7 +53,7 @@ def __init__(self):
         self.add_argument('--eval_percent', type=float, default=1,
                           help='sample some percentage for evaluation.')
 
-        self.add_argument('--gpu', type=int, default=-1,
+        self.add_argument('--gpu', type=int, default=[-1], nargs='+', 
                           help='use GPU')
```
Comment (zheng-da) [inline apps/kg/train.py]: please document how multiple GPUs can be specified.

File: apps/kg/models/general_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/models/general_models.py]: do we still need `args.gpu` to indicate which GPU to run the computation?

**Pull request merged**